src/JWTVerifier.php

```diff
@@ -177,53 +174,35 @@ public function verifyAndDecode($jwt)
         }
 
         try {
-            $head_decoded = $this->decodeTokenSegment($tks[0]);
             $body_decoded = $this->decodeTokenSegment($tks[1]);
         } catch (\DomainException $e) {
             throw new InvalidTokenException('Malformed token.');
         }
 
-        if (! is_object($head_decoded) || ! is_object($body_decoded)) {
-            throw new InvalidTokenException('Malformed token.');
-        }
-
-        if (empty($head_decoded->alg)) {
-            throw new InvalidTokenException('Token algorithm not found');
-        }
-
-        if (! $this->supportsAlg($head_decoded->alg)) {
-            throw new InvalidTokenException('Token algorithm not supported');
-        }
-
         // Validate the token audience, if present.
         if (! empty($body_decoded->aud)) {
             $audience = is_array($body_decoded->aud) ? $body_decoded->aud : [$body_decoded->aud];
             if (! count(array_intersect($audience, $this->valid_audiences))) {
-                $message  = 'Invalid token audience '.implode( ', ', $audience );
-                $message .= '; expected '.implode( ', ', $this->valid_audiences );
-                throw new InvalidTokenException($message);
+                throw new InvalidTokenException('Invalid token audience');
             }
         }
 
-        if ('HS256' === $head_decoded->alg) {
-            $secret = $this->client_secret;
+        if ($this->client_secret) {
+            $secret               = $this->client_secret;
+            $this->supported_algs = [ 'HS256' ];
         } else {
-            if (empty($head_decoded->kid)) {
-                throw new CoreException('Token key ID is missing for RS256 token');
-            }
-
             if (empty($body_decoded->iss) || ! in_array($body_decoded->iss, $this->authorized_iss)) {
-                throw new CoreException('We cannot trust on a token issued by `'.$body_decoded->iss.'`');
+                throw new InvalidTokenException('Invalid token issuer');
             }
 
-            $jwks_url                   = $body_decoded->iss.$this->jwks_path;
-            $secret[$head_decoded->kid] = $this->JWKFetcher->requestJwkX5c($jwks_url, $head_decoded->kid);
+            $jwks_url = $this->authorized_iss[0].$this->jwks_path;
+            $secret   = $this->JWKFetcher->requestCompleteJwks($jwks_url);
```

Comment:
can we rename this to "publicKey" or similar? secret sounds symmetric

Reply:
My point is that class is called "JWKFetcher" and if its purpose is to call the JWKs endpoint and convert that cert hash into a pubkey, it shouldn't be handling HS secrets, only RS keys.